front/lib/api/mcp/run_tool.ts

```diff
@@ -126,14 +128,19 @@ export async function* runToolWithStreaming(
     if (isMCPTimeoutError) {
       errorMessage = `The execution of tool ${action.functionCallName} timed out.`;
 
-      // If the tool should be retried on interrupt, we throw an error so the workflow retries the
-      // `runTool` activity. If the tool should not be retried on interrupt, the error is returned to
-      // the model, to let it decide what to do.
+      // If the tool should not be retried on interrupt, the error is returned
+      // to the model, to let it decide what to do. If the tool should be
+      // retried on interrupt, we throw an error so the workflow retries the
+      // `runTool` activity, unless it's the last attempt.
       const retryPolicy =
         getRetryPolicyFromToolConfiguration(toolConfiguration);
       if (retryPolicy === "retry_on_interrupt") {
-        errorMessage += `Error: ${toolError.message}`;
-        throw new Error(errorMessage, { cause: toolError });
+        const info = Context.current().info;
+        const isLastAttempt = info.attempt === RETRY_ON_INTERRUPT_MAX_ATTEMPTS;
```

Comment:
nit: maybe '>=` to be safe in case we miss an interaction due to some other bug.